subsys/net/lib/mqtt/mqtt_decoder.c

```diff
@@ -205,9 +207,427 @@ int fixed_header_decode(struct buf_ctx *buf, uint8_t *type_and_flags,
+		return -EBADMSG;
+	}
+
+	if (unpack_utf8_str(buf, &temp.value) < 0) {
+		return -EBADMSG;
+	}
+
+	prop_len = (2 * sizeof(uint16_t)) + temp.name.size + temp.value.size;
+	if (*remaining_len < prop_len) {
+		return -EBADMSG;
+	}
+
+	*remaining_len -= prop_len;
+	*prop->found = true;
+
+	for (int i = 0; i < CONFIG_MQTT_USER_PROPERTIES_MAX; i++) {
+		if (user_prop[i].name.utf8 == NULL) {
+			chosen = &user_prop[i];
+			break;
+		}
+	}
+
+	if (chosen == NULL) {
+		NET_DBG("Cannot parse all user properties, ignore excess");
+	} else {
+		memcpy(chosen, &temp, sizeof(struct mqtt_utf8_pair));
+	}
+
+	return 0;
+}
+
+static int properties_decode(struct property_decoder *prop, uint8_t cnt,
+			     struct buf_ctx *buf)
+{
+	uint32_t properties_len;
+	int bytes;
+	int err;
+
+	bytes = unpack_variable_int(buf, &properties_len);
+	if (bytes < 0) {
+		return -EBADMSG;
+	}
+
+	bytes += (int)properties_len;
+
+	while (properties_len > 0) {
+		struct property_decoder *current_prop = NULL;
+		uint8_t type;
+
+		/* Decode property type */
+		err = unpack_uint8(buf, &type);
+		if (err < 0) {
+			return -EBADMSG;
+		}
+
+		properties_len--;
+
+		/* Search if the property is supported in the provided property
+		 * array.
+		 */
+		for (int i = 0; i < cnt; i++) {
+			if (type == prop[i].type) {
+				current_prop = &prop[i];
+			}
+		}
+
+		if (current_prop == NULL) {
+			NET_DBG("Unsupported property %u", type);
+			return -EBADMSG;
+		}
+
+		/* Decode property value. */
+		switch (type) {
+		case MQTT_PROP_SESSION_EXPIRY_INTERVAL:
+		case MQTT_PROP_MAXIMUM_PACKET_SIZE:
+			err = decode_uint32_property(current_prop,
+						     &properties_len, buf);
+			break;
+		case MQTT_PROP_RECEIVE_MAXIMUM:
+		case MQTT_PROP_TOPIC_ALIAS_MAXIMUM:
+		case MQTT_PROP_SERVER_KEEP_ALIVE:
+			err = decode_uint16_property(current_prop,
+						     &properties_len, buf);
+			break;
+		case MQTT_PROP_MAXIMUM_QOS:
+		case MQTT_PROP_RETAIN_AVAILABLE:
+		case MQTT_PROP_WILDCARD_SUBSCRIPTION_AVAILABLE:
+		case MQTT_PROP_SUBSCRIPTION_IDENTIFIER_AVAILABLE:
+		case MQTT_PROP_SHARED_SUBSCRIPTION_AVAILABLE:
+			err = decode_uint8_property(current_prop,
+						    &properties_len, buf);
+			break;
+		case MQTT_PROP_ASSIGNED_CLIENT_IDENTIFIER:
+		case MQTT_PROP_REASON_STRING:
+		case MQTT_PROP_RESPONSE_INFORMATION:
+		case MQTT_PROP_SERVER_REFERENCE:
+		case MQTT_PROP_AUTHENTICATION_METHOD:
+			err = decode_string_property(current_prop,
+						     &properties_len, buf);
+			break;
+		case MQTT_PROP_USER_PROPERTY:
+			err = decode_user_property(current_prop,
+						   &properties_len, buf);
+			break;
+		case MQTT_PROP_AUTHENTICATION_DATA:
+			err = decode_binary_property(current_prop,
+						     &properties_len, buf);
+			break;
+		default:
+			err = -ENOTSUP;
+		}
+
+		if (err < 0) {
+			return -EBADMSG;
+		}
+	}
+
+	return bytes;
+}
+
+static int connack_properties_decode(struct buf_ctx *buf,
+				     struct mqtt_connack_param *param)
+{
+	struct property_decoder prop[] = {
```

Comment:
This looks like a quite large structure, doing a quickly calculation we need 17 * 12 = 204 bytes stack space. But it seems that it is difficult to do this differently. Do we need to adjust any stack size for samples or tests etc because of this?

Reply:
The test suite runs on default stack sizes so perhaps the impact isn't that bad in this particular case.